simpleclient/src/main/java/io/prometheus/client/DoubleAdder.java

```diff
@@ -118,6 +118,37 @@ public void reset() {
         internalReset(0L);
     }
 
+    public void set(double x) {
+        for (;;) {
+            Cell[] as;
+            if ((as = cells) != null) { // have cells
+                if (busy == 0 && casBusy()) {
+                    try {
+                        if (cells == as) { // recheck under lock
+                            // allocate new cells instead of updating existing cells
+                            int n = as.length;
+                            Cell[] rs = new Cell[n];
+                            for (int i = 0; i < n; ++i) {
+                                Cell a = as[i];
+                                if (a != null) // avoid unused cells
+                                    rs[i] = new Cell(0L);
```

Comment:
This could be a lot of allocations.

Reply:
I think we could reuse (steal) the cell if it's `0`:
```
rs[i] = a.value == 0L ? a : new Cell(0L);
```

Let me try to think about the use cases:
* only `set` - `cells` is always `null` - no allocations
* only `inc / dec` - `cells` allocated on contention - few allocations
* mostly `inc / dec` and occasional `set`- `cells` allocated on contention, `cells` reallocated on `set` - some allocations
* mostly `set` and occasional `inc / dec` - `cells` allocated rarely, `cells` reallocated on `set` - some allocations
* mixed `set` and `inc / dec` - `cells` allocated on contention, `cells` reallocated on `set` - a lot of allocations

I'm not sure how to find a balance between the tradeoffs for each use case. 🙁

Reply:
Performance wise we only need to worry about the first two cases, however correctness wise we need to worry about everything.

Reply:
If performance is not that important for mixed `set` and `inc / dec`, then we could just set `cells` to `null` in `set`, and let `inc / dec` reallocate cells on contention.

Reply:
That would be a lot simpler.